### ISSUE TITLE
Lb4

### ISSUE THREAD

**Issue by Homka3310** [Homka3310/LearningMachine#5]
import numpy as np
import pandas as pd
from gensim.models import KeyedVectors
from sklearn.metrics.pairwise import cosine_similarity
import re
import string
from nltk.corpus import stopwords, twitter_samples
from nltk.stem import PorterStemmer
from nltk.tokenize import TweetTokenizer

# Завантаження векторів слів
# Використовуємо бінарні версії для ефективності
lang1_embeddings = KeyedVectors.load_word2vec_format("cc.uk.300.bin", binary=True)
lang2_embeddings = KeyedVectors.load_word2vec_format("cc.en.300.bin", binary=True)

# Функція для завантаження словників
def get_dict(file_name):
    my_file = pd.read_csv(file_name, delimiter=' ', header=None, names=['lang1', 'lang2'])
    l1_to_l2 = {}
    for i, row in my_file.iterrows():
        l1_to_l2[row['lang1']] = row['lang2']
    return l1_to_l2

# Завантаження словників
l1_l2_train = get_dict('uk-en.0-5000.txt')
l1_l2_test = get_dict('uk-en.5000-6500.txt')

print(f"Розмір словника перекладів для тренування: {len(l1_l2_train)}")
print(f"Розмір словника перекладів для тестування: {len(l1_l2_test)}")
def get_matrices(l1_l2, l1_embeddings, l2_embeddings):
    X_l = list()
    Y_l = list()
    l1_set = set(l1_embeddings.index_to_key)
    l2_set = set(l2_embeddings.index_to_key)
    for l1_word, l2_word in l1_l2.items():
        if l1_word in l1_set and l2_word in l2_set:
            X_l.append(l1_embeddings[l1_word])
            Y_l.append(l2_embeddings[l2_word])
    X = np.asarray(X_l)
    Y = np.asarray(Y_l)
    return X, Y

X_train, Y_train = get_matrices(l1_l2_train, lang1_embeddings, lang2_embeddings)
print(f"Форма матриці X_train: {X_train.shape}")
print(f"Форма матриці Y_train: {Y_train.shape}")
def compute_loss(X, Y, R):
    m = X.shape[0]
    diff = np.dot(X, R) - Y
    loss = np.sum(diff**2) / m
    return loss

def compute_gradient(X, Y, R):
    m = X.shape[0]
    gradient = np.dot(X.T, np.dot(X, R) - Y) * (2 / m)
    return gradient

def align_embeddings(X, Y, train_steps=400, learning_rate=0.8):
    np.random.seed(129)
    R = np.random.rand(X.shape[1], X.shape[1])
    for i in range(train_steps):
        if i % 25 == 0:
            print(f"Loss at iteration {i} is: {compute_loss(X, Y, R):.4f}")
        gradient = compute_gradient(X, Y, R)
        R -= learning_rate * gradient
    return R

R_train = align_embeddings(X_train, Y_train, train_steps=400, learning_rate=0.8)
def translate(word, R, l1_embeddings, l2_embeddings):
    if word not in l1_embeddings:
        return "Слово відсутнє у словнику"
    
    l1_emb = l1_embeddings[word]
    l2_emb_pred = np.dot(l1_emb, R)
    
    # Використовуємо функцію з sklearn для ефективності
    similarities = cosine_similarity([l2_emb_pred], l2_embeddings.vectors)[0]
    
    # Знаходимо індекс слова з найбільшою подібністю
    best_match_idx = np.argmax(similarities)
    
    return l2_embeddings.index_to_key[best_match_idx]

# Тестування перекладу
print(f"Переклад слова 'сонце': {translate('сонце', R_train, lang1_embeddings, lang2_embeddings)}")
print(f"Переклад слова 'мова': {translate('мова', R_train, lang1_embeddings, lang2_embeddings)}")
print(f"Переклад слова 'людина': {translate('людина', R_train, lang1_embeddings, lang2_embeddings)}")
def test_vocabulary(X_test, Y_test, R, l1_l2_test, l1_embeddings, l2_embeddings):
    correct = 0
    total = 0
    # Отримуємо слова, що відповідають векторам у X_test
    test_words_l1 = list(l1_l2_test.keys())
    
    for i, word_l1 in enumerate(test_words_l1):
        # Перевіряємо, чи слово є у матриці (було знайдене в embedding)
        if word_l1 in l1_embeddings and l1_l2_test[word_l1] in l2_embeddings:
            pred_l2 = translate(word_l1, R, l1_embeddings, l2_embeddings)
            if pred_l2 == l1_l2_test[word_l1]:
                correct += 1
            total += 1
            
    accuracy = correct / total if total > 0 else 0
    return accuracy

# Створюємо матриці для тестового набору
X_test, Y_test = get_matrices(l1_l2_test, lang1_embeddings, lang2_embeddings)

accuracy_test = test_vocabulary(X_test, Y_test, R_train, l1_l2_test, lang1_embeddings, lang2_embeddings)
print(f"\nТочність перекладу на тестовому наборі: {accuracy_test:.4f}")
# Завантаження твітів
all_positive_tweets = twitter_samples.strings('positive_tweets.json')
all_negative_tweets = twitter_samples.strings('negative_tweets.json')
all_tweets = all_positive_tweets + all_negative_tweets

def process_tweet(tweet):
    stemmer = PorterStemmer()
    stopwords_english = stopwords.words('english')
    tweet = re.sub(r'\$\w*', '', tweet)
    tweet = re.sub(r'^RT[\s]+', '', tweet)
    tweet = re.sub(r'https?:\/\/.*[\r\n]*', '', tweet)
    tweet = re.sub(r'#', '', tweet)
    tokenizer = TweetTokenizer(preserve_case=False, strip_handles=True, reduce_len=True)
    tweet_tokens = tokenizer.tokenize(tweet)
    tweets_clean = []
    for word in tweet_tokens:
        if (word not in stopwords_english and word not in string.punctuation):
            stem_word = stemmer.stem(word)
            tweets_clean.append(stem_word)
    return tweets_clean

def get_tweet_embedding(tweet, embeddings):
    clean_words = process_tweet(tweet)
    if not clean_words:
        return np.zeros(300)
    
    word_vectors = [embeddings[word] for word in clean_words if word in embeddings]
    if not word_vectors:
        return np.zeros(300)
    
    return np.mean(word_vectors, axis=0)

# Створюємо словник з векторами для кожного твіту
tweet_embeddings = {i: get_tweet_embedding(tweet, lang2_embeddings) for i, tweet in enumerate(all_tweets)}

print(f"Кількість оброблених твітів: {len(tweet_embeddings)}")
print(f"Розмірність векторів твітів: {list(tweet_embeddings.values())[0].shape}")
def hash_func(embedding, planes):
    dot_product = np.dot(embedding, planes)
    sign_of_dot_product = (dot_product >= 0).astype(int)
    hash_value = sum(sign_of_dot_product * (2**i) for i, _ in enumerate(sign_of_dot_product))
    return hash_value

def make_hash_table(embeddings, planes):
    hash_table = {}
    for i, embedding in embeddings.items():
        h = hash_func(embedding, planes.T)
        if h not in hash_table:
            hash_table[h] = []
        hash_table[h].append(i)
    return hash_table

def init_lsh(embeddings, n_planes, n_tables):
    dim = list(embeddings.values())[0].shape[0]
    planes_list = [np.random.normal(size=(dim, n_planes)) for _ in range(n_tables)]
    tables = [make_hash_table(embeddings, planes) for planes in planes_list]
    return tables, planes_list

# Ініціалізація LSH
N_PLANES = 10
N_TABLES = 20
tables, planes_list = init_lsh(tweet_embeddings, n_planes=N_PLANES, n_tables=N_TABLES)
def lsh_knn(tweet_id, embedding, tables, planes_list, k=5):
    candidates = set()
    for i, table in enumerate(tables):
        planes = planes_list[i]
        hash_value = hash_func(embedding, planes.T)
        if hash_value in table:
            candidates.update(table[hash_value])
    
    if tweet_id in candidates:
        candidates.remove(tweet_id)
        
    candidate_embeddings_list = [tweet_embeddings[idx] for idx in candidates]
    if not candidate_embeddings_list:
        return []
    
    candidate_embeddings = np.array(candidate_embeddings_list)
    similarities = cosine_similarity([embedding], candidate_embeddings)[0]
    
    # Сортуємо кандидатів за подібністю
    sorted_indices = np.argsort(similarities)[::-1]
    
    # Повертаємо ID k найкращих кандидатів
    top_k_candidates = [list(candidates)[i] for i in sorted_indices[:k]]
    
    return top_k_candidates

# Тестування пошуку
tweet_id = 10
tweet = all_tweets[tweet_id]
embedding = tweet_embeddings[tweet_id]
neighbors = lsh_knn(tweet_id, embedding, tables, planes_list, k=3)

print(f"Запит (твіт №{tweet_id}): {tweet}\n")
print("Знайдені найбільш схожі твіти:")
for neighbor_id in neighbors:
    print(f"  - (Твіт №{neighbor_id}): {all_tweets[neighbor_id]}")
